### ISSUE TITLE
`max_unpool` gives wrong gradient when `indices` has duplicate

### ISSUE THREAD

**Issue by upupyy** [pytorch/pytorch#80204]
### 🐛 Describe the bug

`max_unpool` doesn't consider that the `indices` may have duplicate elements in its gradient formula.

```python
import torch

def get_fn():
    indices = torch.tensor([[[4, 2, 3, 5]]], dtype=torch.int64, device='cuda')
    kernel_size = [1]
    stride = [2]
    def fn(input):
        fn_res = torch.nn.functional.max_unpool1d(input, indices, kernel_size, stride=stride,)
        return fn_res
    return fn
fn = get_fn()
input_tensor = torch.tensor([[[0.6418, 0.0359, 0.6088, 0.9408]]], dtype=torch.float64, device='cuda')

print(input_tensor)
# tensor([[[0.6418, 0.0359, 0.6088, 0.9408]]], dtype=torch.float64)
print(fn(input_tensor))
# tensor([[[0.6418, 0.9408, 0.6088, 0.0000, 0.0000, 0.0000, 0.0000]]],
#       dtype=torch.float64)

try:
    input = input_tensor.clone().detach().requires_grad_()
    torch.autograd.gradcheck(fn, (input), check_sparse_nnz=False, atol=0.01, rtol=0.01, check_forward_ad=False, check_backward_ad=True, check_batched_grad=False)
except Exception as e:
    print(e)

# Jacobian mismatch for output 0 with respect to input 0,
# numerical:tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
#         [0.0000, 1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
#         [0.0000, 0.0000, 1.0000, 0.0000, 0.0000, 0.0000, 0.0000],
#         [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000]],
#        device='cuda:0', dtype=torch.float64)
# analytical:tensor([[1., 0., 0., 0., 0., 0., 0.],
#         [0., 1., 0., 0., 0., 0., 0.],
#         [0., 0., 1., 0., 0., 0., 0.],
#         [0., 1., 0., 0., 0., 0., 0.]], device='cuda:0', dtype=torch.float64)
```

In the above example, the output of `max_unpool` only has 3 elements but the gradient computed by reverse-mode has 4 elements.

I am not sure whether this is a big issue since `max_pool` could not return `indices` with duplicate elements.

### Versions

pytorch: 1.11.0
